src/transformStateWithClones.js

```diff
@@ -7,7 +7,25 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  actions.forEach(action => {
+    if (action.type === 'clear') {
+      currentState = {};
+    } else if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+      action.keysToRemove.forEach(key => {
+        delete currentState[key];
+      });
+    }
```

Comment:
Consider using a switch statement instead of multiple if-else conditions as it will improve readability and is more appropriate when you have a limited and known set of conditions to handle.